File: Design/Design_a_Number_Container_System.cpp
// Problem Name: Design a Number Container System
// Problem Link: https://leetcode.com/problems/design-a-number-container-system/

class NumberContainers {
    unordered_map<int,int> mp1;
    unordered_map<int,set<int>> mp2;
public:
    NumberContainers() {
        
    }
    
    void change(int index, int number) {
        if(mp1.find(index)!=mp1.end()) {
            mp2[mp1[index]].erase(index);
            if(mp2[mp1[index]].size() == 0) mp2.erase(mp1[index]);
        }
        mp1[index] = number;
        mp2[number].insert(index);
    }
    
    int find(int number) {
        if(mp2.find(number)==mp2.end()) return -1;
        return *mp2[number].begin();
    }
};

File: Design/Design_a_Number_Container_System.java
// Problem Name: Design a Number Container System
// Problem Link: https://leetcode.com/problems/design-a-number-container-system/

class NumberContainers {

    private Map<Integer, TreeSet<Integer>> mp2;
    private Map<Integer, Integer> mp1;
    public NumberContainers() {
        mp1 = new HashMap<>();
        mp2 = new HashMap<>();
    }
    
    public void change(int index, int number) {
        if(mp1.containsKey(index)) {
            int prevNumber = mp1.get(index);
            mp2.get(prevNumber).remove(index);
            if(mp2.get(prevNumber).isEmpty()) {
                mp2.remove(prevNumber);
            }
        }
        mp1.put(index, number);
        mp2.putIfAbsent(number, new TreeSet<>());
        mp2.get(number).add(index);
    }
    
    public int find(int number) {
        if(mp2.containsKey(number)) {
            return mp2.get(number).first();
        }
        return -1;
    }
}
